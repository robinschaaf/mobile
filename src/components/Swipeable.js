import React, { Component } from 'react'
import {
  Animated,
  PanResponder,
  Platform,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import theme from '../theme'
import { connect } from 'react-redux'
import StyledText from './StyledText'
import SwipeSubject from './SwipeSubject'
import { clamp } from 'ramda'
import { setNextSubject } from '../actions/subject'

//needs to be absolutely positioned below the question, otherwise it covers up
//everything above and prevents it from being touchable
const toTop = (Platform.OS === 'ios') ? 135 : 125
const SWIPE_THRESHOLD = 90
const leftOverlayColor = theme.swipeLeft
const rightOverlayColor = theme.swipeRight

const mapStateToProps = (state, ownProps) => ({
  workflow: state.classifier.workflow[ownProps.workflowID] || {},
  subject: state.classifier.subject[ownProps.workflowID] || {},
  classification: state.classifier.classification[ownProps.workflowID] || {},
  annotations: state.classifier.annotations[ownProps.workflowID] || {},
  subjectSizes: state.classifier.subjectSizes[ownProps.workflowID] || {},
  seenThisSession: state.classifier.seenThisSession[ownProps.workflowID] || [],
  questionContainerHeight: state.classifier.questionContainerHeight[ownProps.workflowID] || 0,
})

const mapDispatchToProps = (dispatch) => ({
  setNextSubject() {
    dispatch(setNextSubject())
  },
})

export class Swipeable extends Component {
  constructor(props) {
    super(props)
    this.state = {
      pan: new Animated.ValueXY(),
      enterAnim: new Animated.Value(.9),
    }
  }

  isSwipeGesture(gestureState) {
    return gestureState.dx > 5 || gestureState.dx < -5
  }

  componentWillMount() {
    this._panResponder = PanResponder.create({
    onStartShouldSetResponder: (evt, gestureState) => this.isSwipeGesture(gestureState),
    onMoveShouldSetPanResponder: (evt, gestureState) => this.isSwipeGesture(gestureState),

    onPanResponderGrant: () => {
      this.state.pan.setOffset({x: this.state.pan.x._value, y: this.state.pan.y._value})
      this.state.pan.setValue({x: 0, y: 0})
    },

    onPanResponderMove: Animated.event([
      null, {dx: this.state.pan.x, dy: this.state.pan.y + 100},
    ]),

    onPanResponderRelease: (e, {vx, vy}) => {
      this.state.pan.flattenOffset()
      let velocity

      if (vx >= 0) {
        velocity = clamp(3, 4, vx)
      } else if (vx < 0) {
        velocity = clamp(3, 4, vx * -1) * -1
      }

      if (Math.abs(this.state.pan.x._value) > SWIPE_THRESHOLD) {
        //negative=left (no = index 1)
        //positive=right (yes = index 0)
        const answer = ( this.state.pan.x._value < 0 ? 1 : 0 )
        Animated.decay(this.state.pan, {
          velocity: {x: velocity, y: vy},
          deceleration: 0.98
        }).start(this.props.onAnswered(answer))
      } else {
        Animated.spring(this.state.pan, {
          toValue: {x: 0, y: 0},
          friction: 4
        }).start()
      }
    }
    })
  }

  render() {
    const imageSizeStyle = { width: this.props.subjectSizes.resizedWidth, height: this.props.subjectSizes.resizedHeight }
    const swipeableSize = { width: this.props.subjectSizes.resizedWidth, height: this.props.subjectSizes.resizedHeight + 10 }
    let pan = this.state.pan

    let [translateX, translateY] = [pan.x, pan.y]

    let rotate = pan.x.interpolate({inputRange: [-200, 0, 200], outputRange: ['-30deg', '0deg', '30deg']})
    let opacityRight = pan.x.interpolate({inputRange: [0, 100, 200], outputRange: [0, .6, .8]})
    let opacityLeft = pan.x.interpolate({inputRange: [-200, -100, 0], outputRange: [.8, .6, 0]})
    let opacityRightText = pan.x.interpolate({inputRange: [0, 30], outputRange: [0, 1]})
    let opacityLeftText = pan.x.interpolate({inputRange: [-30, 0], outputRange: [1, 0]})

    let animatedCardStyles = {transform: [{translateX}, {translateY}, {rotate}]}
    let leftOverlayStyle = {backgroundColor: leftOverlayColor, opacity: opacityLeft}
    let rightOverlayStyle = {backgroundColor: rightOverlayColor, opacity: opacityRight}
    let leftOverlayTextStyle = {opacity: opacityLeftText}
    let rightOverlayTextStyle = {opacity: opacityRightText}

    return (
      <View style={[styles.container, {top: toTop + this.props.questionContainerHeight}]}>
        <View style={styles.subjectContainer}>
          <Animated.View
            style={[styles.imageContainer, animatedCardStyles, swipeableSize]}
            {...this._panResponder.panHandlers}>

            <View style={swipeableSize}>
              <SwipeSubject
                inFront={true}
                subject={this.props.subject}
                subjectSizes={this.props.subjectSizes}
                seenThisSession={this.props.seenThisSession}
                setNextSubject={this.props.setNextSubject}
              />
              <Animated.View style={[styles.overlayContainer, leftOverlayStyle, imageSizeStyle]} />
              <Animated.View style={[styles.overlayContainer, leftOverlayTextStyle, imageSizeStyle]}>
                <StyledText additionalStyles={[styles.answerOverlayText]} text={ this.props.answers[0].label } />
              </Animated.View>

              <Animated.View style={[styles.overlayContainer, rightOverlayStyle, imageSizeStyle]} />
              <Animated.View style={[styles.overlayContainer, rightOverlayTextStyle, imageSizeStyle]}>
                <StyledText additionalStyles={[styles.answerOverlayText]} text={ this.props.answers[1].label } />
              </Animated.View>
            </View>
          </Animated.View>
        </View>
      </View>
    )
  }
}

const styles = EStyleSheet.create({
  container: {
    position: 'absolute',
    right: 0,
    left: 0,
    bottom: 0,
    backgroundColor: 'transparent',
    flex: 1,
    paddingVertical: 10,
    paddingHorizontal: 20,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  subjectContainer: {
    position: 'absolute',
    top: 0,
    right: 0,
    left: 0,
    bottom: 0,
    elevation: 0,
    backgroundColor: 'transparent',
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  overlayContainer: {
    borderRadius: 2,
    backgroundColor: 'transparent',
    position: 'absolute',
    top: 20,
    right: 0,
    left: 0,
    bottom: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
  answerOverlayText: {
    fontSize: 30,
    color: 'white'
  },
})

Swipeable.propTypes = {
  subject: React.PropTypes.shape({
    id: React.PropTypes.string
  }),
  nextSubject: React.PropTypes.shape({
    id: React.PropTypes.string
  }),
  subjectSizes: React.PropTypes.object,
  seenThisSession: React.PropTypes.arrayOf(React.PropTypes.string),
  workflow: React.PropTypes.shape({
    first_task: React.PropTypes.string,
    tasks: React.PropTypes.object,
  }),
  onAnswered: React.PropTypes.func,
  questionContainerHeight: React.PropTypes.number,
  setNextSubject: React.PropTypes.func,
  answers: React.PropTypes.arrayOf(React.PropTypes.object)
}

export default connect(mapStateToProps, mapDispatchToProps)(Swipeable)
