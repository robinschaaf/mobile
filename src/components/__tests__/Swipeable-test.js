import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
jest.mock('../SwipeSubject', () => 'SwipeSubject')

import { Swipeable } from '../Swipeable'

const workflow = {
  first_task: 'T0',
  tasks: {
    T0: {
      question: 'Is it?',
      answers: [ { label: 'Yes' }, { label: 'No' } ],
    }
  }
}
const subjectSizes = {
  resizedWidth: 100,
  resizedHeight: 100
}
const subject = {
    id: '3424'
}
const nextSubject = {
    id: '234'
}

const answers=[
  { label: 'No' },
  { label: 'Yes' },
]

it('renders correctly', () => {
  const tree = renderer.create(
    <Swipeable
      saveAnnotation={jest.fn}
      saveThenStartNewClassification={jest.fn}
      workflow={workflow}
      subjectSizes={subjectSizes}
      subject={subject}
      nextSubject={nextSubject}
      onAnswered={jest.fn}
      answers={answers}
    />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
