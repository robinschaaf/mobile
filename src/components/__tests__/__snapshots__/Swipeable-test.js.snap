exports[`test renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "top": NaN,
      },
    ]
  }>
  <View
    style={undefined}>
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "height": 110,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
            Object {
              "rotate": "0deg",
            },
          ],
          "width": 100,
        }
      }>
      <TouchableOpacity
        activeOpacity={0.8}
        onPress={undefined}
        style={
          Object {
            "height": 110,
            "width": 100,
          }
        }>
        <SwipeSubject
          inFront={true}
          seenThisSession={undefined}
          setNextSubject={undefined}
          subject={
            Object {
              "id": "3424",
            }
          }
          subjectSizes={
            Object {
              "resizedHeight": 100,
              "resizedWidth": 100,
            }
          } />
        <View
          style={
            Object {
              "backgroundColor": "#E45950",
              "height": 100,
              "opacity": 0,
              "width": 100,
            }
          } />
        <View
          style={
            Object {
              "height": 100,
              "opacity": 0,
              "width": 100,
            }
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                undefined,
                Array [
                  undefined,
                ],
              ]
            }>
            No
          </Text>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#00979D",
              "height": 100,
              "opacity": 0,
              "width": 100,
            }
          } />
        <View
          style={
            Object {
              "height": 100,
              "opacity": 0,
              "width": 100,
            }
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                undefined,
                Array [
                  undefined,
                ],
              ]
            }>
            Yes
          </Text>
        </View>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
